### PR TITLE
python deep analysis

### PR DESCRIPTION
**Pull request by mahigangal**
## 📝 Description

This PR adds a complete analysis module for Python projects, along with an accompanying test suite. The analyzer is designed to extract meaningful insights from a repository’s Python source code and interpret what those artifacts reveal about the author’s skills, design decisions, and understanding of key Computer Science concepts.

1. Object-Oriented Design Analysis
- Detects classes, methods, inheritance relationships, and overridden methods.
- Identifies __init__ presence, dunder methods, and internal state via attribute assignments.
- Tracks encapsulation through _private attributes.
- Computes a weighted OOP Score (0–1) with a qualitative rating and explanation.

2. Data Structure Usage Analysis
- Counts list/dict/set/tuple literals and comprehensions.
- Detects use of specialized structures (defaultdict, Counter).
- Interprets the developer’s data structure choices and patterns.

3. Algorithmic & Complexity Analysis
- Counts functions and identifies nested loop depth (inferring potential O(n²) behaviour).
- Detects the use of tools such as sorted(), heapq, and bisect.
- Summarizes complexity awareness and algorithmic reasoning.

4. Narrative Insight Generation
- A human-readable narrative is generated for: Object-oriented reasoning, Data structure selection, Complexity and algorithmic behaviour
- These narratives translate raw metrics into meaningful analysis that aligns with capstone expectations for deep interpretation of code artifacts, not just surface-level statistics.

5. Graceful Error Handling
- Syntax-error files are recorded without interrupting the analysis.
- Analyzer works even on empty or non-Python directories.

**Closes:** #201 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

The tests collectively confirm correctness of the metrics, stability across edge cases, correct construction of the narrative field and integration of all components into the final output.

All tests pass locally.
---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

Screenshot 2025-11-30 at 12 36 36 AM

Screenshot 2025-11-30 at 12 37 02 AM

Screenshot 2025-11-30 at 12 37 57 AM

</details>

### CODE REVIEW

File: src/python_oop_metrics.py [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+        print("\n[OOP]")
+        print(narrative["oop"])
+    if "data_structures" in narrative:
+        print("\n[Data Structures]")
+        print(narrative["data_structures"])
+    if "complexity" in narrative:
+        print("\n[Complexity & Algorithms]")
+        print(narrative["complexity"])
+
+    print("\n" + "="*60 + "\n")
```
Comment (ThunderIW) [inline src/python_oop_metrics.py]: Good return statement to return the pretty format of this good Job

File: src/python_oop_metrics.py
Comment (ThunderIW) [inline src/python_oop_metrics.py]: Returns the correct information when I test it with my tinyscripts folder, Really good job on this and something we really needed for our system 

Image

**Review by ThunderIW — APPROVED** [review]
Really good Coding here @mahigangal, everything works as intended, good job on doing the non-LLM approach for grabbing the deeper analysis. PR is approved and I have left you some comments and if you have concerns let me know
